### ISSUE TITLE
Various type-related issues

### ISSUE THREAD

**Issue by simplexx** [ConsenSys/mythril#67]
I installed mythril from the git code. When I use "python3 myth --init-db", I get the following error:
  File ".../mythril/mythril/ether/ethcontract.py", line 18, in __init__
    code = re.sub(r'(_+[A-Za-z0-9]+_+)', 'aa' * 20, code)
  File "/usr/lib/python3.5/re.py", line 182, in sub
    return _compile(pattern, flags).sub(repl, string, count)
TypeError: cannot use a string pattern on a bytes-like object

Fixing this with a small change:
old: code = re.sub(r'(_+[A-Za-z0-9]+_+)', 'aa' * 20, code)
new: code = re.sub(b'(_+[A-Za-z0-9]+_+)', 'aa' * 20, code)

makes the db init work (at least it indexes the contracts..)

But searching the contracts does not work as well: 

sudo python3 myth --search "func#demo(address)#" leads to the following error:

File ".../mythril/mythril/ether/util.py", line 14, in safe_decode
    if (hex_encoded_string.startswith("0x")):
TypeError: startswith first arg must be bytes or a tuple of bytes, not str

And it goes on with new issues, after fixing an issue.

I am not a python dev. so I really don't know where the problem could be. All the problems seem to be Types related and it seems like only the files in the mythril/mythril/ether/ folder have these issues.

**Comment by b-mueller** [comment]
Seems like a mixup between `str` and `bytes` types. Can you please provide the output of:

```
$ python --version
$ pip freeze
```

Also, what OS are you working on?

**Comment by simplexx** [comment]
I tried it on ubuntu v16.04.3 as well as v17.10.1 andhad the same issues on both. On v16.04.3 I used python 3.5.

$ python3 --version
 Python 3.6.3
$ pip3.6 freeze
apturl==0.5.2
asn1crypto==0.22.0
attrdict==2.0.0
Brlapi==0.6.5
BTrees==4.4.1
certifi==2017.4.17
chardet==3.0.4
command-not-found==0.3
cryptography==1.9
cupshelpers==1.0
cytoolz==0.9.0
defer==1.0.6
distro-info==0.17
eth-abi==1.0.0
eth-account==0.1.0a2
eth-hash==0.1.0
eth-keyfile==0.5.1
eth-keys==0.2.0b3
eth-rlp==0.1.0a2
eth-utils==1.0.1
ethereum==2.3.0
future==0.16.0
hexbytes==0.1.0b1
httplib2==0.9.2
idna==2.5
keyring==10.4.0
keyrings.alt==2.2
language-selector==0.1
laser-ethereum==0.4.4
launchpadlib==1.10.5
lazr.restfulclient==0.13.5
lazr.uri==1.0.3
louis==3.0.0
lru-dict==1.1.6
Mako==1.0.7
MarkupSafe==1.0
mythril==0.12.5
oauth==1.0.1
olefile==0.44
pbkdf2==1.3
persistent==4.2.4.2
pexpect==4.2.1
Pillow==4.1.1
py-ecc==1.4.2
py-solc==2.1.0
pyandoc==0.2.0
pycrypto==2.6.1
pycryptodome==3.4.11
pycups==1.9.73
pyethash==0.1.27
pygobject==3.24.1
pysha3==1.0.2
python-apt==1.4.0b3
python-debian==0.1.30
pyxdg==0.25
PyYAML==3.12
reportlab==3.4.0
repoze.lru==0.7
requests==2.18.4
rlp==0.6.0
scrypt==0.8.6
SecretStorage==2.3.1
semantic-version==2.6.0
simplejson==3.11.1
six==1.10.0
system-service==0.3
systemd-python==234
toolz==0.9.0
transaction==2.2.0
ubuntu-drivers-common==0.0.0
ufw==0.35
unattended-upgrades==0.1
urllib3==1.21.1
usb-creator==0.3.3
wadllib==1.3.2
web3==4.0.0b11
xkit==0.0.0
z3-solver==4.5.1.0.post2
zc.lockfile==1.2.1
ZConfig==3.2.0
ZODB==5.3.0
zodbpickle==1.0
zope.interface==4.3.2

**Comment by b-mueller** [comment]
My guess it that there were some changes to the Ethereum-related packages that break the existing code. 

Does manually in installing the prior versions fix it? Try:

```
$ pip install eth-abi==0.5.0 eth-keyfile==0.4.1 eth-keys==0.1.0b4 eth-tester==0.1.0b11 eth-utils==0.7.4
```

**Comment by simplexx** [comment]
The lowest versions I was able to use (due to min. version requirements) were:

eth-utils-1.0.0b1
eth-abi-1.0.0-beta.1
eth-keys-0.2.0b2 (b1 is actually allowed, but throws an error)
eth-keyfile-0.5.0

Unfortunately, the issue still persists.

I found some info that could be helpful:
from here: https://docs.python.org/3/howto/pyporting.html
"The constructors of both str and bytes have different semantics for the same arguments between Python 2 & 3. Passing an integer to bytes in Python 2 will give you the string representation of the integer: bytes(3) == '3'. But in Python 3, an integer argument to bytes will give you a bytes object as long as the integer specified, filled with null bytes: bytes(3) == b'\x00\x00\x00'. A similar worry is necessary when passing a bytes object to str. In Python 2 you just get the bytes object back: str(b'3') == b'3'. But in Python 3 you get the string representation of the bytes object: str(b'3') == "b'3'""

after some research it seems to me that the files in the folder "mythril/mythril/ether/" are still python 2.x and need to be ported to 3.x in order for the project to work on 3.5. (For example missing "b" prefix for literal that presents binary data, missing proper string <-> byte conversions etc.).
I could be wrong though, but I i tested it on 2 different fresh installs of ubuntu with different python versions (3.5, 3.6) and both had the same issues.

$sudo myth --init-db
Resuming synchronization from block 5173100
Traceback (most recent call last):
  File "/usr/local/bin/myth", line 4, in <module>
    __import__('pkg_resources').run_script('mythril==0.12.5', 'myth')
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 742, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1503, in run_script
    exec(code, namespace, namespace)
  File "/usr/local/lib/python3.6/dist-packages/mythril-0.12.5-py3.6.egg/EGG-INFO/scripts/myth", line 235, in <module>
    contract_storage.initialize(eth, args.sync_all)
  File "/usr/local/lib/python3.6/dist-packages/mythril-0.12.5-py3.6.egg/mythril/ether/contractstorage.py", line 95, in initialize
    code = ETHContract(contract_code, tx['input'])
  File "/usr/local/lib/python3.6/dist-packages/mythril-0.12.5-py3.6.egg/mythril/ether/ethcontract.py", line 18, in __init__
    code = re.sub(r'(_+[A-Za-z0-9]+_+)', 'aa' * 20, code)
  File "/usr/lib/python3.6/re.py", line 191, in sub
    return _compile(pattern, flags).sub(repl, string, count)
TypeError: cannot use a string pattern on a bytes-like object

**Comment by b-mueller** [comment]
Hmm Mythril written specifically for Python 3 so I don't think it's related to Python version.

My suspicion is that some of the `eth_` functions, like `eth_getCode`, in the Ethereum packages are now returning byte arrays instead of strings. This would explain how things suddenly break in a lot of places.

I think we'll need to lock package versions in setup.py more tightly.

Trying to set up Mythril in a fresh environment now to check if I get the same errors!

**Comment by simplexx** [comment]
Thanks a lot for your help! I managed to get it to work. The important part was to install from pypi first and then run the setup from the repo in order to be able to work with the code. Also some packages had to be installed manually.

**Comment by b-mueller** [comment]
Awesome! What packages were missing?

I'll keep this issue open for now because I suspect other users might run into similar problems.

**Comment by simplexx** [comment]
In the pypi installation,
-laser-ethereum==0.4.4 was not installed automatically, not sure why. I installed it with
"pip3.6 install laser-ethereum==0.4.4 --user" without any issues.
-eth-utils==0.7.* is required, but a newer version (0.8.1) was automatically installed, so I had to downgrade manually 
Also I had an issue where the mythril installation from pip did not work anymore. Tried everything, deleted everything and tried again etc. - but the problem persisted. The only thing that helped was installing mythril from the git repo.
